das/ipfs_storage_service.go

```diff
@@ -0,0 +1,159 @@
+	}
+	return &IpfsStorageService{
+		ipfsHelper: ipfsHelper,
+		ipfsApi:    ipfsHelper.GetAPI(),
+	}, nil
+}
+
+func hashToCid(hash common.Hash) (cid.Cid, error) {
+	multiEncodedHashBytes, err := multihash.Encode(hash[:], multihash.KECCAK_256)
+	if err != nil {
+		return cid.Cid{}, err
+	}
+
+	_, multiHash, err := multihash.MHFromBytes(multiEncodedHashBytes)
+	if err != nil {
+		return cid.Cid{}, err
+	}
+
+	return cid.NewCidV1(cid.Raw, multiHash), nil
+}
+
+// GetByHash retrieves and reconstructs one batch's data, using IPFS to retrieve the preimages
+// for each chunk of data and the dastree nodes.
+func (s *IpfsStorageService) GetByHash(ctx context.Context, hash common.Hash) ([]byte, error) {
+	oracle := func(h common.Hash) ([]byte, error) {
+		thisCid, err := hashToCid(h)
+		if err != nil {
+			return nil, err
+		}
+
+		ipfsPath := path.IpfsPath(thisCid)
+		log.Trace("Retrieving IPFS path", "path", ipfsPath.String())
+
+		rdr, err := s.ipfsApi.Block().Get(ctx, ipfsPath)
+		if err != nil {
+			return nil, err
+		}
+
+		data, err := ioutil.ReadAll(rdr)
+		if err != nil {
+			return nil, err
+		}
+		return data, nil
+	}
+
+	return dastree.Content(hash, oracle)
+}
+
+// Put stores all the preimages required to reconstruct the dastree for single batch,
+// ie the hashed data chunks and dastree nodes.
+// This takes advantage of IPFS supporting keccak256 on raw data blocks for calculating
+// its CIDs, and the fact that the dastree structure uses keccak256 for addressing its
+// nodes, to directly store the dastree structure in IPFS.
+// IPFS default block size is 256KB and dastree max block size is 64KB so each dastree
+// node and data chunk easily fits within an IPFS block.
+func (s *IpfsStorageService) Put(ctx context.Context, data []byte, _ uint64) error {
+	record := func(_ common.Hash, value []byte) error {
+		blockStat, err := s.ipfsApi.Block().Put(
+			ctx,
+			bytes.NewReader(value),
+			options.Block.CidCodec("raw"), // Store the data in raw form since the hash in the CID must be the hash
+			// of the preimage for our lookup scheme to work.
+			options.Block.Hash(multihash.KECCAK_256, -1), // Use keccak256 to calculate the hash to put in the block's
+			// CID, since it is the same algo used by dastree.
+			options.Block.Pin(true)) // Keep the data in the local IPFS repo, don't GC it.
+		if err != nil {
+			return err
+		}
+		log.Trace("Wrote IPFS path", "path", blockStat.Path().String())
+		return nil
+	}
+
+	_, err := dastree.RecordHash(record, data)
+	return err
+}
+
+func (s *IpfsStorageService) ExpirationPolicy(ctx context.Context) (arbstate.ExpirationPolicy, error) {
+	return arbstate.KeepForever, nil
+}
+
+func (s *IpfsStorageService) Sync(ctx context.Context) error {
+	return nil
+}
+func (s *IpfsStorageService) Close(ctx context.Context) error {
+	return s.ipfsHelper.Close()
+}
+func (s *IpfsStorageService) String() string {
+	return "IpfsStorageService"
+}
+func (s *IpfsStorageService) HealthCheck(ctx context.Context) error {
```

Comment:
Is this file `gofmt`'d ?

Reply:
Discussed offline, this is formatted correctly. My mistake!

Reply:
Yeah I have it set to auto-format when I save. I added the whitespace between the functions which gofmt doesn't do.